File: Sources/Helper/LRErrors.m
//
//  LRErrors.m
//  LoginRadiusSDK
//
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//

#import "LRErrors.h"
#import "LoginRadiusError.h"

@implementation LRErrors

+ (NSError*)serviceCancelled:(NSString*)action {
	return [NSError errorWithCode:LRErrorCodeRaaSCancelled
					  description:@"User Registration Service cancelled"
					failureReason:[NSString stringWithFormat:@"User registration with action: %@ cancelled", action]];
}

+ (NSError*)userRegistraionFailed {
	return [NSError errorWithCode:LRErrorCodeRaaSUserRegistrationFailed
					  description:@"User registration failed"
					failureReason:@"User registration failed with no status"];
}

+ (NSError*)userLoginFailed {
	return [NSError errorWithCode:LRErrorCodeRaaSUserLoginFailed
					  description:@"User login failed"
					failureReason:@"User login failed since token not recieved"];
}

+ (NSError*)userForgotPasswordFailed {
	return  [NSError errorWithCode:LRErrorCodeRaaSUserForgotPasswordFailed
					   description:@"User forgot password failed"
					 failureReason:@"User forgot password failed with no status"];
}

+ (NSError*)userIsNotVerified:(NSString*)field {
	return  [NSError errorWithCode:LRErrorCodeUserIsNotVerified
					   description:[NSString stringWithFormat:@"User %@ is not verified", field]
					 failureReason:[NSString stringWithFormat:@"User need to verify %@ in order to proceed", field]];
}

+ (NSError*)userEmailVerificationFailed {
	return  [NSError errorWithCode:LRErrorCodeRaaSUserEmailVerificationFailed
					   description:@"User email verification failed"
					 failureReason:@"User email verification failed with no status"];
}

+ (NSError*)userSocialLoginFailed {
	return  [NSError errorWithCode:LRErrorCodeRaaSUserSocialLoginFailed
					   description:@"User social login failed"
					 failureReason:@"User social login failed since token not recieved"];
}

+ (NSError*)userResetPasswordFailed {
	return  [NSError errorWithCode:LRErrorCodeRaaSUserResetPasswordFailed
					   description:@"User reset password failed"
					 failureReason:@"User reset password failed failed with no status"];

}

+ (NSError*)tokenEmpty {
	return [NSError errorWithCode:LRErrorCodeAccessTokenEmpty
					  description:@"User profile fetching failed"
					failureReason:@"Access token is empty or null"];
}

+ (NSError*)userProfieWithErrorCode:(NSString*)errorCode {
	return [NSError errorWithCode:LRErrorCodeUserProfileError
					  description:@"User profile error"
					failureReason:[NSString stringWithFormat:@"User profile error with error code %@", errorCode]];
}

+ (NSError*)userProfileError {
	return [NSError errorWithCode:LRErrorCodeUserProfileError
			   description:@"User profile error"
			 failureReason:@"User profile is either bloked or returned with an errorCode"];
}

+ (NSError*)userProfileRequireAdditionalFields {
	return [NSError errorWithCode:LRErrorCodeUserRequireAdditionalFieldsError
			   description:@"User profile has incomplete required fields"
			 failureReason:@"User profile has null fields that are required before registration/loggin in"];
}

+ (NSError*)socialLoginFetchFailed {
	return [NSError errorWithCode:LRErrorCodeWebSocialLoginFetchFailed
					  description:@"Failed to fetch social login list"
					failureReason:@"Failed to fetch social login list from server"];
}

+ (NSError*)socialLoginCancelled:(NSString*) provider {
	return [NSError errorWithCode:LRErrorCodeWebSocialLoginCancelled
					  description:@"Social Login cancelled"
					failureReason:[NSString stringWithFormat:@"Social login with %@ is cancelled", provider]];
}

+ (NSError*)socialLoginFailed:(NSString*) provider {
	return [NSError errorWithCode:LRErrorCodeWebSocialLoginFailed
								description:@"Social Login failed"
							  failureReason:[NSString stringWithFormat:@"Social login with %@ is failed since token is not recieved", provider]];
}

+ (NSError*)nativeTwitterLoginNoAccount {
    return [NSError errorWithCode:LRErrorCodeNativeTwitterLoginCancelled
                      description:@"Twitter login not available"
                    failureReason:@"Native Twitter login is not available since user doesn't have a configured twitter account"];
}

+ (NSError*)nativeTwitterLoginCancelled {
	return [NSError errorWithCode:LRErrorCodeNativeTwitterLoginNotAvailable
					  description:@"Twitter login cancelled"
					failureReason:@"Native twitter login is cancelled by user"];
}

+ (NSError*)nativeTwitterLoginFailed {
	return [NSError errorWithCode:LRErrorCodeNativeTwitterLoginFailed
					  description:@"Twitter login failed"
					failureReason:@"Native twitter login is not granted by user"];
}

+ (NSError*)nativeFacebookLoginCancelled {
	return [NSError errorWithCode:LRErrorCodeNativeFacebookLoginCancelled
					  description:@"Facebook Login cancelled"
					failureReason:@"Faceook native login is cancelled"];

}

+ (NSError*)nativeFacebookLoginFailed {
	return [NSError errorWithCode:LRErrorCodeNativeFacebookLoginFailed 
					  description:@"Facebook login failed"
					failureReason:@"Your app should only ask for read permissions for first time login"];
}

+ (NSError*)nativeFacebookLoginFailedMixedPermissions {
	return [NSError errorWithCode:LRErrorCodeNativeFacebookLoginFailed
					  description:@"Facebook login failed"
					failureReason:@"Your app can't ask for both read and write permissions"];
}

+ (NSError*)touchIDNotAvailable {
    return [NSError errorWithCode:LRErrorCodeTouchIDNotAvailable
                      description:@"Touch ID authentication failed"
                    failureReason:@"The User's device cannot be authenticated using TouchID"];
}

+ (NSError*)touchIDNotDeviceOwner {
    return [NSError errorWithCode:LRErrorCodeTouchIDNotAvailable
                      description:@"Touch ID authentication failed"
                    failureReason:@"TouchID Authentiction failed since the user is not the device's owner"];
}

+ (NSError*)faceIDNotAvailable {
    return [NSError errorWithCode:LRErrorCodeFaceIDNotAvailable
                      description:@"Face ID authentication failed"
                    failureReason:@"The User's device cannot be authenticated using FaceID"];
}

+ (NSError*)faceIDNotDeviceOwner {
    return [NSError errorWithCode:LRErrorCodeFaceIDNotAvailable
                      description:@"Face ID authentication failed"
                    failureReason:@"FaceID Authentiction failed since the user is not the device's owner"];
}
@end

File: Sources/LRErrors.swift
//
//  LRErrors.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//

import Foundation


public class LRErrors {
    public static func serviceCancelled(action: String) -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSCancelled.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User Registration Service cancelled",
                NSLocalizedFailureReasonErrorKey: "User registration with action: \(action) cancelled"
            ]
        )
    }
    
    public static func userRegistraionFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSUserRegistrationFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User registration failed",
                NSLocalizedFailureReasonErrorKey: "User registration failed with no status"
            ]
        )
    }
    
    public static func userLoginFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSUserLoginFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User login failed",
                NSLocalizedFailureReasonErrorKey: "User login failed since token not received"
            ]
        )
    }
    
    public static func userForgotPasswordFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSUserForgotPasswordFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User forgot password failed",
                NSLocalizedFailureReasonErrorKey: "User forgot password failed with no status"
            ]
        )
    }
    
    public static func userIsNotVerified(field: String) -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.UserIsNotVerified.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User \(field) is not verified",
                NSLocalizedFailureReasonErrorKey: "User need to verify \(field) in order to proceed"
            ]
        )
    }
    
    public static func userEmailVerificationFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSUserEmailVerificationFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User email verification failed",
                NSLocalizedFailureReasonErrorKey: "User email verification failed with no status"
            ]
        )
    }
    
    public static func userSocialLoginFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSUserSocialLoginFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User social login failed",
                NSLocalizedFailureReasonErrorKey: "User social login failed since token not received"
            ]
        )
    }
    
    public static func userResetPasswordFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.RaaSUserResetPasswordFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User reset password failed",
                NSLocalizedFailureReasonErrorKey: "User reset password failed with no status"
            ]
        )
    }
    
    public static func tokenEmpty() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.AccessTokenEmpty.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User profile fetching failed",
                NSLocalizedFailureReasonErrorKey: "Access token is empty or null"
            ]
        )
    }
    
    public static func userProfieWithErrorCode(errorCode: String) -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.UserProfileError.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User profile error",
                NSLocalizedFailureReasonErrorKey: "User profile error with error code \(errorCode)"
            ]
        )
    }
    
    public static func userProfileError() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.UserProfileError.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User profile error",
                NSLocalizedFailureReasonErrorKey: "User profile is either blocked or returned with an errorCode"
            ]
        )
    }
    
    public static func userProfileRequireAdditionalFields() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.UserRequireAdditionalFieldsError.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "User profile has incomplete required fields",
                NSLocalizedFailureReasonErrorKey: "User profile has null fields that are required before registration/logging in"
            ]
        )
    }
    
    public static func socialLoginFetchFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.WebSocialLoginFetchFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Failed to fetch social login list",
                NSLocalizedFailureReasonErrorKey: "Failed to fetch social login list from the server"
            ]
        )
    }
    
    public static func socialLoginCancelled(provider: String) -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.WebSocialLoginCancelled.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Social Login cancelled",
                NSLocalizedFailureReasonErrorKey: "Social login with \(provider) is cancelled"
            ]
        )
    }
    
    public static func socialLoginFailed(provider: String) -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.WebSocialLoginFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Social Login failed",
                NSLocalizedFailureReasonErrorKey: "Social login with \(provider) is failed since the token is not received"
            ]
        )
    }
    
    public static func nativeTwitterLoginNoAccount() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.NativeTwitterLoginCancelled.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Twitter login not available",
                NSLocalizedFailureReasonErrorKey: "Native Twitter login is not available since the user doesn't have a configured Twitter account"
            ]
        )
    }
    
    public static func nativeTwitterLoginCancelled() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.NativeTwitterLoginNotAvailable.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Twitter login cancelled",
                NSLocalizedFailureReasonErrorKey: "Native Twitter login is cancelled by the user"
            ]
        )
    }
    
    public static func nativeTwitterLoginFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.NativeTwitterLoginFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Twitter login failed",
                NSLocalizedFailureReasonErrorKey: "Native Twitter login is not granted by the user"
            ]
        )
    }
    
    public static func nativeFacebookLoginCancelled() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.NativeFacebookLoginCancelled.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Facebook Login cancelled",
                NSLocalizedFailureReasonErrorKey: "Facebook native login is cancelled"
            ]
        )
    }
    
    public static func nativeFacebookLoginFailed() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.NativeFacebookLoginFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Facebook login failed",
                NSLocalizedFailureReasonErrorKey: "Your app should only ask for read permissions for first-time login"
            ]
        )
    }
    
    public static func nativeFacebookLoginFailedMixedPermissions() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.NativeFacebookLoginFailed.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Facebook login failed",
                NSLocalizedFailureReasonErrorKey: "Your app can't ask for both read and write permissions"
            ]
        )
    }
    
    public static func touchIDNotAvailable() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.TouchIDNotAvailable.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Touch ID authentication failed",
                NSLocalizedFailureReasonErrorKey: "The user's device cannot be authenticated using TouchID"
            ]
        )
    }
    
    public static func touchIDNotDeviceOwner() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.TouchIDNotAvailable.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Touch ID authentication failed",
                NSLocalizedFailureReasonErrorKey: "TouchID authentication failed since the user is not the device's owner"
            ]
        )
    }
    
    public static func faceIDNotAvailable() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.FaceIDNotAvailable.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Face ID authentication failed",
                NSLocalizedFailureReasonErrorKey: "The user's device cannot be authenticated using FaceID"
            ]
        )
    }
    
    public static func faceIDNotDeviceOwner() -> NSError {
        return NSError(
            domain: "LoginRadiusErrorDomain",
            code: LRErrorCode.FaceIDNotAvailable.rawValue,
            userInfo: [
                NSLocalizedDescriptionKey: "Face ID authentication failed",
                NSLocalizedFailureReasonErrorKey: "FaceID authentication failed since the user is not the device's owner"
            ]
        )
    }
}

